### ISSUE TITLE
Handling of casts to nullable type differs between 2.1.0-preview2 and 2.0.0

### ISSUE THREAD

**Issue by mbashov** [aspnet/EntityFrameworkCore#11690]
```c#
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.EntityFrameworkCore;

class Program
{
    static void Main(string[] args)
    {
        using (var db = new Context())
        {
            db.Database.EnsureDeleted();
            db.Database.EnsureCreated();

            db.Entities.Add(new Main());
            db.SaveChanges();

            var result = db
                .Entities
                .Select(e => new { DependentId = (int?)e.Dependent.FirstOrDefault().Id })
                .First();
            Console.WriteLine(result.DependentId?.ToString() ?? "null");
        }
    }
}

public class Context : DbContext
{
    public DbSet<Main> Entities { get; set; }
    public DbSet<Dependent> Values { get; set; }
}

public class Main
{
    public int Id { get; set; }
    public ICollection<Dependent> Dependent { get; set; }
}

public class Dependent
{
    public int Id { get; set; }
    public int MainId { get; set; }
    public Main Main { get; set; }
}

```

This query returns `null` on EF Core 2.0.2, but `0` on 2.1.0-preview2. Is this breaking change intentional?

**Comment by maumar** [comment]
The change was intentional and is a result of fixing https://github.com/aspnet/EntityFrameworkCore/issues/10294

**Comment by mbashov** [comment]
This is really unfortunate. Now we're forced to write something like

```c#
            var result = db
                .Entities
                .Select(e => new
                {
                    FirstDependentCreated = e.Dependent.Any()
                        ? e.Dependent.Select(d => d.Created).OrderBy(d => d).FirstOrDefault()
                        : (DateTime?)null,
                })
                .First();
```

And the generated SQL is long and suboptimal, making two scans on the dependent table:

```sql
      SELECT CASE
          WHEN (
              SELECT CASE
                  WHEN EXISTS (
                      SELECT 1
                      FROM "Values" AS "v"
                      WHERE "e"."Id" = "v"."MainId")
                  THEN CAST(1 AS BIT) ELSE CAST(0 AS BIT)
              END
          ) = 1
          THEN COALESCE((
              SELECT "d"."Created"
              FROM "Values" AS "d"
              WHERE "e"."Id" = "d"."MainId"
              ORDER BY "d"."Created"
              LIMIT 1
          ), '0001-01-01 00:00:00') ELSE NULL
      END AS "FirstDependentCreated"
      FROM "Entities" AS "e"
      LIMIT 1
```

For comparison, 2.0 version generates this nice and simple query:
```sql
      SELECT (
          SELECT "d"."Created"
          FROM "Values" AS "d"
          WHERE "e"."Id" = "d"."MainId"
          ORDER BY "d"."Created"
          LIMIT 1
      ) AS "FirstDependentCreated"
      FROM "Entities" AS "e"
      LIMIT 1
```

**Comment by ajcvickers** [comment]
Clearing milestone to get a better understanding of this in triage.

**Comment by mbashov** [comment]
@maumar Indeed, this works, and looks like it covers all cases where a single field from a dependent entity is selected. Thanks for the tip.

**Comment by maumar** [comment]
fixed in 542de5520a48b5acd6d6219f73675984059daaa0

**Comment by maumar** [comment]
@mbashov we ended up reverting the behavior to the 2.0, i.e.

`customers.Select(c => (int?)c.Orders.Where(o => false).FirstOrDefault().Id)` will now return collection of nulls

however:

`customers.Select(c => c.Orders.Where(o => false).Select(o => o.Id).FirstOrDefault())` will produce collection of 0s
